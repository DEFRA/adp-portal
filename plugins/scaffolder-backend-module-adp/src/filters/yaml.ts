import type { JsonValue } from '@backstage/types';
import YAML from 'yaml';

export function yaml(value: JsonValue) {
  return YAML.stringify(value).slice(0, -1);
}
