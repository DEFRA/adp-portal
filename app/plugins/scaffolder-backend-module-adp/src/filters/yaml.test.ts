import type { JsonValue } from '@backstage/types';
import { yaml } from './yaml';

const inputs: JsonValue[] = [
  123,
  'abc',
  true,
  false,
  null,
  `This is some multiline text
and this is line 2!`,
  `This input: contains special characters! "|'{}[],&*#?<>=%@/\\
And even a newline!`,
  '|-\n  This already looks like yaml but shouldnt be treated as yaml',
  {
    some: {
      object: {
        with: ['values'],
      },
    },
  },
];

interface TestCase {
  readonly input: JsonValue;
  readonly inputJSON: string;
}

describe('yaml', () => {
  it.each<TestCase>(
    inputs.map(input => (
      { input, inputJSON: JSON.stringify(input) }
    )),
  )(
    'Should correctly handle $inputJSON',
    ({ input }) => {
      // act
      const actual = yaml(input);

      // assert
      expect(actual).toMatchSnapshot();
    },
  );
});
