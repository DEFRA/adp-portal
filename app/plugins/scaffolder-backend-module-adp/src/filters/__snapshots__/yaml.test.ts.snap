// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml Should correctly handle "|-\\n  This already looks like yaml but shouldnt be treated as yaml" 1`] = `
"|-
|-
  This already looks like yaml but shouldnt be treated as yaml"
`;

exports[`yaml Should correctly handle "This input: contains special characters! \\"|'{}[],&*#?<>=%@/\\\\\\nAnd even a newline!" 1`] = `
"|-
This input: contains special characters! "|'{}[],&*#?<>=%@/\\
And even a newline!"
`;

exports[`yaml Should correctly handle "This is some multiline text\\nand this is line 2!" 1`] = `
"|-
This is some multiline text
and this is line 2!"
`;

exports[`yaml Should correctly handle "abc" 1`] = `"abc"`;

exports[`yaml Should correctly handle {"some":{"object":{"with":["values"]}}} 1`] = `
"some:
  object:
    with:
      - values"
`;

exports[`yaml Should correctly handle 123 1`] = `"123"`;

exports[`yaml Should correctly handle false 1`] = `"false"`;

exports[`yaml Should correctly handle null 1`] = `"null"`;

exports[`yaml Should correctly handle true 1`] = `"true"`;
